src/main/java/org/eclipse/yasson/internal/deserializer/types/EnumDeserializer.java

```diff
@@ -13,21 +13,51 @@
 package org.eclipse.yasson.internal.deserializer.types;
 
 import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
 
 import org.eclipse.yasson.internal.DeserializationContextImpl;
+import org.eclipse.yasson.internal.model.ClassModel;
+import org.eclipse.yasson.internal.model.PropertyModel;
 
 /**
  * Deserializer of the {@link Enum}.
  */
 class EnumDeserializer extends TypeDeserializer {
 
+    private final Map<String, ? extends Enum<?>> nameToConstantMap;
+
     EnumDeserializer(TypeDeserializerBuilder builder) {
         super(builder);
+
+        nameToConstantMap = createNameToConstantMap(builder);
+    }
+
+    private static <E extends Enum<E>> Map<String, E> createNameToConstantMap(TypeDeserializerBuilder builder) {
+        Map<String, E> nameToConstantMap = null;
+        Class<?> clazz = builder.getClazz();
+
+        if (clazz.isEnum()) {
+            try {
+                @SuppressWarnings("unchecked")
+                Class<E> enumClazz = (Class<E>) clazz;
+                nameToConstantMap = new HashMap<>();
+                ClassModel classModel = builder.getJsonbContext().getMappingContext().getOrCreateClassModel(clazz);
+
+                for (E enumConstant : enumClazz.getEnumConstants()) {
+                    PropertyModel model = classModel.getPropertyModel(enumConstant.name());
+                    nameToConstantMap.put(model.getReadName(), enumConstant);
+                }
+            } catch (ClassCastException classCastException) {
+                throw new IllegalArgumentException("EnumDeserializer can only be used with Enum types");
+            }
+        }
+        return nameToConstantMap;
     }
 
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({"unchecked", "rawtypes"})
     @Override
     Object deserializeStringValue(String value, DeserializationContextImpl context, Type rType) {
-        return Enum.valueOf((Class<Enum>) rType, value);
+        return nameToConstantMap == null ? Enum.valueOf((Class<Enum>) rType, value) : nameToConstantMap.get(value);
```

Comment:
I like this with the result caching. Maybe just a thought, when can it be null?

Reply:
Yes, caching makes this faster and this map shouldn't be big and take a long time for creation because enums are normally not so long.

As about the null check: right now, the de- or serializer constructor is called only once with the check if the class is an enum. But in the constructor itself it just a class, so we don't know is this an enum class or not. I wrote a code to prevent the NPE in case the class isn't an enum class somehow. So this why the cache map could be null.

In such case, we would have an CCE in this line, even if I prevent a NPE above. The only scenario where this wouldn't happen is when we have another type here than the class we have in a builder in method above. Which is just another kind of error constellation.

So if you think it is OK to have some exception here - I could do this. I just semi-automatically wrote a secure code with fallback to the old logic in some edge case.